### PR TITLE
Bosonic backend - backend.py and states.py

### PR DESCRIPTION
**Pull request by elib20**
**Context:** Adds bosonic backend files and associated tests

**Description of the Change:**
- Creates a backend.py file for the bosonic backend with all gaussian operations (non-gaussian to be added next PR)
- Creates a states.py file for the bosonic backend with tailored methods
- Modifies the conftest.py, backend init, and backend base file to add the bosonic backend
- Modifies relevant preexisting tests to test the bosonic backend
- Adds new tests to check unique features of bosonic backend
- Removes some tests that will be re-included once the bosonic engine is added

**Benefits:**
- building the bosonic backend infrastructure

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: tests/backend/test_multimode_state_preparations.py
Comment (nquesada) [inline tests/backend/test_multimode_state_preparations.py:625]: missing endline

File: tests/backend/test_states.py
Comment (nquesada) [inline tests/backend/test_states.py:64]:
```suggestion
```

File: tests/backend/test_multimode_state_preparations.py
Comment (nquesada) [inline tests/backend/test_multimode_state_preparations.py:625]:
```suggestion
        )

```

File: tests/backend/test_states.py
Comment (nquesada) [inline tests/backend/test_states.py:658]:
```suggestion
        assert np.allclose(dm,dm_compare)

```

File: strawberryfields/backends/base.py
Comment (josh146) [inline strawberryfields/backends/base.py:695]: @elib20 just out of curiosity, is a separate `BaseBosonic` class required, or could we simply use `BaseGaussian`? Asking because this class looks very similar to `BaseGaussian`, but I imagine that might change as more PRs get merged!

File: strawberryfields/backends/states.py
Comment (josh146) [inline strawberryfields/backends/states.py:1476]:
```suggestion
          array[float]: an array of shape ``(num_weights, 2N)``
```
Also, ``N`` is not defined here?

File: strawberryfields/backends/states.py
Comment (josh146) [inline strawberryfields/backends/states.py:1484]: Might also be good to use 'of shape ``(a, b, c)`` nomenclature here, and define the `N` :)

File: strawberryfields/backends/states.py
Comment (josh146) [inline strawberryfields/backends/states.py:1509]: Should this be `self._hbar`? It might be important to use the value of hbar set at state creation, rather than the global hbar value.

Because the global hbar value might change _after_ this state was created!

File: strawberryfields/backends/states.py
Comment (josh146) [inline strawberryfields/backends/states.py:1596]:
```suggestion
                "The number of specified modes cannot be larger than the number of subsystems."
```

File: tests/backend/test_states_probabilities.py
Comment (josh146) [inline tests/backend/test_states_probabilities.py:70]:
```suggestion
@pytest.mark.backends("fock", "tf", "gaussian")
```

is this not supported on the bosonic backend?

File: tests/integration/test_algorithms.py
Comment (josh146) [inline tests/integration/test_algorithms.py:22]:
```suggestion
@pytest.mark.backends("fock","tf", "gaussian")
```

**Review by josh146 — COMMENTED** [review]
Left a few comments @elib20, but overall looks like amazing work. This was a fly-by review, so no need to wait for my approval - approval from the other tagged reviewers will be fine :)

File: strawberryfields/backends/base.py
Comment (elib20) [reply strawberryfields/backends/base.py:695]: We actually don't add anything to the BaseBosonic class in future PRs, although we certainly add new things to the BosonicBackend class. My understanding was that we made a separate BaseBosonic class as it was safer, for future iterations, to have something separate in case there are any unique modifications beyond the BaseGaussian class.

File: strawberryfields/backends/base.py
Comment (elib20) [inline strawberryfields/backends/base.py:723]:
```suggestion
            r (array): vector of means in :math:`(x_1,p_1,x_2,p_2,\dots)` ordering
            V (array): covariance matrix in :math:`(x_1,p_1,x_2,p_2,\dots)` ordering
```

File: strawberryfields/backends/states.py
Comment (elib20) [inline strawberryfields/backends/states.py:1473]:
```suggestion
        r"""The vectors of means describing the Bosonic state on N modes.
```

File: strawberryfields/backends/states.py
Comment (elib20) [reply strawberryfields/backends/states.py:1476]:
```suggestion
          array[complex]: an array of shape ``(num_weights, 2N)``
```

File: strawberryfields/backends/states.py
Comment (elib20) [reply strawberryfields/backends/states.py:1484]:
```suggestion
          array[complex]: an array of shape ``(num_weights, 2N, 2N)``
```

File: strawberryfields/backends/states.py
Comment (elib20) [reply strawberryfields/backends/states.py:1509]:
```suggestion
        pur *= self._hbar ** self.num_modes
```

File: strawberryfields/backends/states.py
Comment (elib20) [inline strawberryfields/backends/states.py:1481]:
```suggestion
        r"""The covariance matrices describing the Bosonic state on N modes.
```

File: tests/backend/test_states_probabilities.py
Comment (elib20) [reply tests/backend/test_states_probabilities.py:70]: Not at the moment. The main focus so far for the backend's use so far has been GKP states, where you only need Gaussian measurements. We plan to roll out Fock measurements as a future improvement.

File: strawberryfields/backends/base.py
Comment (elib20) [reply strawberryfields/backends/base.py:695]: I suppose a minor difference that is reflected in one of the docstrings now is that the bosonic backend uses a different quadrature ordering convention that the Gaussian backend.

Comment (josh146) [reply strawberryfields/backends/base.py:695]: In that case, I would inherit from `BaseGaussian` here, and _only_ overwrite the method that has the docstring change :)

This will reduce unneeded duplication, and help make more manageable keeping the documentation up to date.

Comment (josh146) [reply strawberryfields/backends/base.py:695]: (there's generally no issue having backends 'share' a base class -- for example, the tf and fock backends both share/inherit from `BaseFock`.)

Comment (elib20) [reply strawberryfields/backends/base.py:695]: I guess the init method is also slightly different, but I see your point. I will have it inherit from BaseGaussian 👍

Comment (josh146) [reply strawberryfields/backends/base.py:695]: Ah, in that case feel free to also overwrite init (and you can call `super().__init__(*args, **kwargs)` to reduce duplication)

Comment (elib20) [reply strawberryfields/backends/base.py]: Done. Thanks!

File: strawberryfields/backends/base.py
Comment (antalszava) [inline strawberryfields/backends/base.py:693]: Unless explicitly needed at the moment, let's add this in when adding the compiler too so that it won't be forgotten that this is specified here.
```suggestion
```

File: strawberryfields/backends/base.py
Comment (antalszava) [inline strawberryfields/backends/base.py:697]: Seems like a new option to be added in the docstring of the `BaseBackend.supports` method.

Also, would `bosonic` be the best way to describe operations that the bosonic backend is capable of applying? (E.g., could `gaussian_linear_comb` or something else describe the distinct operations better?)

File: strawberryfields/backends/__init__.py
Comment (antalszava) [inline strawberryfields/backends/__init__.py:79]: The module docstring above will have to be changed so that the new bosonic features are included in the documentation. This is important because Sphinx directly uses those references to create the documentation.

It is also worth checking that e.g., the new ``BaseBosonic`` class appears in the list of backends, the docstring of functions render correctly, etc. For this, the `docs/readthedocs.org:strawberryfields` CI check is really helpful, as it offers a "new version" of the SF documentation - so worth checking out that all's fine there!

**Review by antalszava — COMMENTED** [review]
Hi @elib20, as we are changing to code that is user-facing, we'll need to consider that every docstring will serve the purpose of creating the documentation that is hosted on the website. Therefore, docstrings (e.g., module docstrings) need to be extended to include the new bosonic classes, methods, functions.

Left a couple of preliminary comments for now, will continue later today, ping me if anything comes up. 🙂

File: strawberryfields/backends/base.py
Comment (elib20) [reply strawberryfields/backends/base.py:695]: Will modify

File: strawberryfields/backends/__init__.py
Comment (elib20) [reply strawberryfields/backends/__init__.py:79]: Will modify

File: strawberryfields/backends/bosonicbackend/backend.py
Comment (elib20) [inline strawberryfields/backends/bosonicbackend/backend.py:26]:
```suggestion
    """Bosonic backend class for simulating states that can be represented as 
    linear combinations of Gaussian functions in phase space."""
```

**Review by thisac — COMMENTED** [review]
Awesome work @elib20! :tada: I've mainly looked at the code quality, and haven't looked at any tests yet. Some methods I just think need a bit of empty lines to make them a bit more legible (although this is fairly minor, and feel free to do as you prefer). 

I also had some comments on potentially unnecessary parts due to the class inheritance, that might be good to check. Otherwise this seems like a massive undertaking, and a great achievement. :fire:

File: strawberryfields/backends/bosonicbackend/backend.py
Comment (nquesada) [reply strawberryfields/backends/bosonicbackend/backend.py:202]: Because the Gaussian backend uses internally a completely different convention that anywhere else.

**Review by antalszava — COMMENTED** [review]
Looking really great @elib20! 💯 Left some suggestions, but no major blockers. Will go for another iteration once the comments from are addressed.

File: strawberryfields/backends/states.py
Comment (josh146) [reply strawberryfields/backends/states.py:1624]: :+1: Best to inherit docstrings where possible, as it avoids duplication and helps with maintenance.

File: strawberryfields/backends/base.py
Comment (elib20) [reply strawberryfields/backends/base.py:701]:
```suggestion
        r"""Prepare a Gaussian state. Note the different basis-ordering from the 
        GaussianBackend.
```

File: strawberryfields/backends/bosonicbackend/backend.py
Comment (elib20) [inline strawberryfields/backends/bosonicbackend/backend.py:48]:
```suggestion
        r"""Adds new modes to the circuit each with a number of Gaussian peaks
```

File: strawberryfields/backends/bosonicbackend/backend.py
Comment (elib20) [reply strawberryfields/backends/bosonicbackend/backend.py:43]: Will delete this and many other docstrings since they can be inherited 👍

File: strawberryfields/backends/bosonicbackend/backend.py
Comment (elib20) [reply strawberryfields/backends/bosonicbackend/backend.py:246]: If you use measure_homodyne, you can either sample a value (using the circuit.homodyne method) or supply the outcome directly (using the circuit.post_select_homodyne). In both cases, however, you also need to update the other modes.

circuit.post_select_generaldyne is the method that performs the update on the other modes.

When you use the circuit.homodyne method, a sample is returned and circuit.post_select_generaldyne is automatically called within circuit.homodyne to update the other modes

When you supply the post-selection value, really all you need is to update the other modes, since you already have the measurement outcome. So when circuit.post_select_homodyne gets called here, it applies circuit.post_select_generaldyne, and the value you passed to it, which is already known, can just directly be used as the returned value of measure_homodyne.

File: strawberryfields/backends/bosonicbackend/backend.py
Comment (elib20) [reply strawberryfields/backends/bosonicbackend/backend.py:185]: I hope with the next PR it will become clear why it was done this way. Perhaps we can decide then whether to rewrite?

File: strawberryfields/backends/bosonicbackend/backend.py
Comment (elib20) [reply strawberryfields/backends/bosonicbackend/backend.py:360]: When a user deletes all modes, the arrays becomes np.array([]), whereas for methods in the states.py file to work, we need np.array([[]]).

File: strawberryfields/backends/bosonicbackend/backend.py
Comment (elib20) [reply strawberryfields/backends/bosonicbackend/backend.py:84]: For vacuum, we just reset it to one weight. You'll see in the next PR, that for complicated states, we have an init_circuit method that writes all the correct weights.

File: strawberryfields/backends/states.py
Comment (elib20) [reply strawberryfields/backends/states.py:1449]:
```suggestion
            other (:class:`~.BaseBosonicState`): ``BaseBosonicState`` to compare against

```

Comment (elib20) [reply strawberryfields/backends/states.py:1449]: Good catch!

File: strawberryfields/backends/states.py
Comment (elib20) [reply strawberryfields/backends/states.py:1452]: Fundamentally, two physical states are the same if their Wigner functions are the same. However, you can have different mathematical expressions (representations) that yield the same Wigner function. This just checks that the mathematical expressions are the same, not that the states are the same, since to check the states is harder to do.

File: strawberryfields/backends/states.py
Comment (elib20) [inline strawberryfields/backends/states.py:1497]:
```suggestion
        r"""Yields the purity of the state by calculating the integral over the Wigner function
        multiplied with itself.
        
        Returns
                float: purity of the state
        """
```

File: strawberryfields/backends/states.py
Comment (elib20) [reply strawberryfields/backends/states.py:1738]: yes

File: tests/backend/test_multimode_state_preparations.py
Comment (elib20) [reply tests/backend/test_multimode_state_preparations.py:551]: As I said for this entire class of tests, it is a direct copy of the test for the Gaussian backend, the only change being that the basis ordering is different, and that BaseBosonicState has the method .covs instead of Gaussian .cov. If we change this one, we would also need to change the Gaussian backend test, which has been just fine.

File: tests/backend/test_states.py
Comment (elib20) [reply tests/backend/test_states.py:624]:
```suggestion
        """Test weights are correct for a Gaussian state represented using the bosonic backend."""
```

Comment (elib20) [reply tests/backend/test_states.py:624]: The test is just for a Gaussian state in this case. We will have analogous tests for non-Gaussian states, but for the moment only Gaussian states can be created with this PR.

File: tests/backend/test_states.py
Comment (elib20) [inline tests/backend/test_states.py:633]:
```suggestion
        """Test purities are correct for Gaussian states represented using the bosonic backend."""
```

File: strawberryfields/backends/states.py
Comment (elib20) [reply strawberryfields/backends/states.py:1643]: I tried deleting them, but then I got this error when I ran the tests: 
> TypeError: Can't instantiate abstract class BaseBosonicState with abstract methods all_fock_probs, fidelity, ket, number_expectation, poly_quad_expectation

File: strawberryfields/backends/bosonicbackend/backend.py
Comment (elib20) [inline strawberryfields/backends/bosonicbackend/backend.py:200]:
```suggestion
        raise NotImplementedError("Bosonic backend does not yet support Fock measurements.")
```

File: strawberryfields/backends/bosonicbackend/backend.py
Comment (elib20) [inline strawberryfields/backends/bosonicbackend/backend.py:203]:
```suggestion
        raise NotImplementedError("Bosonic backend does not yet support threshold measurements.")
```

File: strawberryfields/backends/states.py
Comment (elib20) [reply strawberryfields/backends/states.py:1734]: It is done now!
```suggestion
```

File: tests/backend/test_states.py
Comment (antalszava) [inline tests/backend/test_states.py:622]:
```suggestion
    """Tests the bosonic state class."""
```

File: tests/integration/test_measurement_integration.py
Comment (antalszava) [inline tests/integration/test_measurement_integration.py:66]:
```suggestion

    @pytest.mark.backends("fock","tf","gaussian")
```

File: tests/integration/test_ops_integration.py
Comment (antalszava) [inline tests/integration/test_ops_integration.py:72]:
```suggestion

    @pytest.mark.backends("fock","tf","gaussian")
```

**Review by antalszava — APPROVED** [review]
Hi @elib20, awesome addition! 🥇 Looks good to me, left a couple of comments, once they are addressed this is merge-ready from my side.

File: strawberryfields/backends/states.py
Comment (elib20) [reply strawberryfields/backends/states.py:1452]: When you have just one weight, as is the case for Gaussian states, the state is uniquely defined by the mean and covariance, so it is sufficient to just check that those are equal for two Gaussian states.

File: strawberryfields/backends/states.py
Comment (elib20) [reply strawberryfields/backends/states.py:1421]: done

File: strawberryfields/backends/bosonicbackend/backend.py
Comment (elib20) [reply strawberryfields/backends/bosonicbackend/backend.py]: done

File: strawberryfields/backends/states.py
Comment (elib20) [reply strawberryfields/backends/states.py]: done

File: tests/backend/test_multimode_state_preparations.py
Comment (nquesada) [reply tests/backend/test_multimode_state_preparations.py:551]: Done. The tests are now split, at the expense of some lines repeated lines.

**Review by nquesada — APPROVED** [review]

File: strawberryfields/backends/bosonicbackend/backend.py
Comment (nquesada) [reply strawberryfields/backends/bosonicbackend/backend.py:244]: the representations of the `Bosonic` backend always have one more dimension or index than the corresponding ones in the `Gaussian` backend.

**Pull request merged**
